src/diffusers/models/autoencoders/dc_ae.py

```diff
@@ -0,0 +1,877 @@
+        current_block_type = block_type[d] if isinstance(block_type, list) else block_type
+        block = build_block(
+            block_type=current_block_type,
+            in_channels=width if d > 0 else input_width,
+            out_channels=width,
+            norm=norm,
+            act=act,
+        )
+        stage.append(block)
+    return stage
+
+
+def build_downsample_block(block_type: str, in_channels: int, out_channels: int, shortcut: Optional[str]) -> nn.Module:
+    if block_type == "Conv":
+        block = nn.Conv2d(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=3,
+            stride=2,
+            padding=1,
+        )
+    elif block_type == "ConvPixelUnshuffle":
+        block = ConvPixelUnshuffleDownsample2D(
+            in_channels=in_channels, out_channels=out_channels, kernel_size=3, factor=2
+        )
+    else:
+        raise ValueError(f"block_type {block_type} is not supported for downsampling")
+    if shortcut is None:
+        pass
+    elif shortcut == "averaging":
+        shortcut_block = PixelUnshuffleChannelAveragingDownsample2D(
+            in_channels=in_channels, out_channels=out_channels, factor=2
+        )
+        block = ResidualBlock(block, shortcut_block)
+    else:
+        raise ValueError(f"shortcut {shortcut} is not supported for downsample")
+    return block
+
+
+def build_upsample_block(block_type: str, in_channels: int, out_channels: int, shortcut: Optional[str]) -> nn.Module:
+    if block_type == "ConvPixelShuffle":
+        block = ConvPixelShuffleUpsample2D(
+            in_channels=in_channels, out_channels=out_channels, kernel_size=3, factor=2
+        )
+    elif block_type == "InterpolateConv":
+        block = Upsample2D(channels=in_channels, use_conv=True, out_channels=out_channels)
+    else:
+        raise ValueError(f"block_type {block_type} is not supported for upsampling")
+    if shortcut is None:
+        pass
+    elif shortcut == "duplicating":
+        shortcut_block = ChannelDuplicatingPixelUnshuffleUpsample2D(
+            in_channels=in_channels, out_channels=out_channels, factor=2
+        )
+        block = ResidualBlock(block, shortcut_block)
+    else:
+        raise ValueError(f"shortcut {shortcut} is not supported for upsample")
+    return block
+
+
+def build_encoder_project_in_block(in_channels: int, out_channels: int, factor: int, downsample_block_type: str):
+    if factor == 1:
+        block = nn.Conv2d(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=3,
+            padding=1,
+        )
+    elif factor == 2:
+        block = build_downsample_block(
+            block_type=downsample_block_type, in_channels=in_channels, out_channels=out_channels, shortcut=None
+        )
+    else:
+        raise ValueError(f"downsample factor {factor} is not supported for encoder project in")
+    return block
+
+
+def build_encoder_project_out_block(
```

Comment:
let's remove this and create these layers directly in the `__init__` method of `Encoder`

Reply:
done